In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        if ".pkl" in filename:
            print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S14/S14.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S11/S11.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S13/S13.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S10/S10.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S8/S8.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S5/S5.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S7/S7.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S9/S9.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S15/S15.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S2/S2.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S6/S6.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S3/S3.pkl
/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD

In [2]:
# /kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S14/S14.pkl
import pandas as pd

# Pfad zur Datei
file_path = '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S14/S14.pkl'

# Einlesen der Pickle-Datei
with open(file_path, 'rb') as file:
    data = pd.read_pickle(file)

# Anzeigen der Struktur (z. B. Keys, falls es ein Dictionary ist)
print(type(data))
print(data.keys())


<class 'dict'>
dict_keys(['signal', 'label', 'subject'])


In [3]:
import numpy as np
import collections

# Sensorsignale und Labels extrahieren
signals = data['signal']  # dict mit 'eda', 'ecg', etc.
labels = data['label']    # NumPy-Array (z. B. [1 1 1 2 2 2 ...])

print(signals.keys())
# Beispiel: EDA-Signal


# Filter: nur Zeilen mit label == 2
#type(data[labels == 2])
#cells_with_stress = data[labels == 2]
#type(cells_with_stress)
# Sortieren: z. B. nach EDA-Werten (aufsteigend)
#cells_with_stress_sorted = np.sort(cells_with_stress, axis=0)

#print("EDA-Signale mit Label=2, sortiert:\n", cells_with_stress_sorted)

import collections

# Häufigkeiten zählen
label_counts = dict(collections.Counter(labels))

# Label-Bedeutungen laut WESAD (inkl. erweiterter Interpretation)
label_meanings = {
    0: "Keine Annotation / Übergang",
    1: "Baseline (Ruhe)",
    2: "Stressphase (z. B. Rechenaufgabe)",
    3: "Amusement (z. B. lustiges Video)",
    4: "Entspannung / Meditation (optional)",
    5: "Fragebogen / Pause (optional)",
    6: "Vorbereitung / Setup (optional)",
    7: "Technischer Übergang / unbekannt (optional)"
}

# Übersicht ausgeben
for label, count in sorted(label_counts.items()):
    meaning = label_meanings.get(label, "Unbekannt")
    print(f"Label {label}: {count} Einträge – {meaning}")

print(data['subject'])

dict_keys(['chest', 'wrist'])
Label 0: 1648499 Einträge – Keine Annotation / Übergang
Label 1: 826000 Einträge – Baseline (Ruhe)
Label 2: 472500 Einträge – Stressphase (z. B. Rechenaufgabe)
Label 3: 260401 Einträge – Amusement (z. B. lustiges Video)
Label 4: 555800 Einträge – Entspannung / Meditation (optional)
Label 5: 44100 Einträge – Fragebogen / Pause (optional)
Label 6: 38500 Einträge – Vorbereitung / Setup (optional)
Label 7: 37800 Einträge – Technischer Übergang / unbekannt (optional)
S14


In [4]:
data['signal']['wrist']

{'ACC': array([[ -23.,   31.,  127.],
        [-128.,  -31.,   62.],
        [ -54.,  -58.,   26.],
        ...,
        [ -47.,  -45.,    6.],
        [ -47.,  -44.,    8.],
        [ -47.,  -45.,    8.]]),
 'BVP': array([[  -8.27],
        [  -9.52],
        [ -10.68],
        ...,
        [-123.97],
        [-160.08],
        [-195.94]]),
 'EDA': array([[0.277864],
        [0.266351],
        [0.195992],
        ...,
        [0.080859],
        [0.060391],
        [0.175257]]),
 'TEMP': array([[31.97],
        [31.97],
        [31.97],
        ...,
        [31.21],
        [31.21],
        [31.21]])}

## Daten sortieren wo nur Stress = 2 ist

Es sind viel zu wenige Daten für Stress. Oben sind es nich 470 Tsd unten nur noch 2700?!

In [5]:
from scipy.signal import resample


labels = data['label']  # Länge z. B. 3883600
eda = data['signal']['wrist']['EDA']  # Länge z. B. 22192

print(f"Label length: {len(labels)} | EDA length: {len(eda)}")

# Labels auf die Länge der EDA-Daten bringen
labels_resampled = resample(labels.astype(float), len(eda))
labels_resampled = np.round(labels_resampled).astype(int)

# Maske anwenden
mask = labels_resampled == 2
eda_label2 = eda[mask]

# In DataFrame
df_eda_label2 = pd.DataFrame({
    'EDA': eda_label2[:, 0],
    'Label': [2] * len(eda_label2)
})

print(df_eda_label2)


Label length: 3883600 | EDA length: 22192
           EDA  Label
0     0.339268      2
1     0.337989      2
2     0.336710      2
3     0.337989      2
4     0.337989      2
...        ...    ...
2697  0.336710      2
2698  0.335430      2
2699  0.316242      2
2700  0.314962      2
2701  0.286819      2

[2702 rows x 2 columns]


### Baseline

In [6]:
from scipy.signal import resample


labels = data['label']  # Länge z. B. 3883600
eda = data['signal']['wrist']['EDA']  # Länge z. B. 22192

print(f"Label length: {len(labels)} | EDA length: {len(eda)}")

# Labels auf die Länge der EDA-Daten bringen
labels_resampled = resample(labels.astype(float), len(eda))
labels_resampled = np.round(labels_resampled).astype(int)

# Maske anwenden
mask = labels_resampled == 1
eda_label2 = eda[mask]

# In DataFrame
df_eda_label2 = pd.DataFrame({
    'EDA': eda_label2[:, 0],
    'Label': [1] * len(eda_label2)
})

print(df_eda_label2)


Label length: 3883600 | EDA length: 22192
           EDA  Label
0     0.303449      1
1     0.295773      1
2     0.291936      1
3     0.293215      1
4     0.285539      1
...        ...    ...
4721  0.339268      1
4722  0.335430      1
4723  0.321359      1
4724  0.313683      1
4725  0.321359      1

[4726 rows x 2 columns]


### Amusement

In [7]:
from scipy.signal import resample


labels = data['label']  # Länge z. B. 3883600
eda = data['signal']['wrist']['EDA']  # Länge z. B. 22192

print(f"Label length: {len(labels)} | EDA length: {len(eda)}")

# Labels auf die Länge der EDA-Daten bringen
labels_resampled = resample(labels.astype(float), len(eda))
labels_resampled = np.round(labels_resampled).astype(int)

# Maske anwenden
mask = labels_resampled == 3
eda_label2 = eda[mask]

# In DataFrame
df_eda_label2 = pd.DataFrame({
    'EDA': eda_label2[:, 0],
    'Label': [3] * len(eda_label2)
})

print(df_eda_label2)


Label length: 3883600 | EDA length: 22192
           EDA  Label
0     0.303449      3
1     0.314962      3
2     0.316242      3
3     0.317521      3
4     0.316242      3
...        ...    ...
1484  0.313683      3
1485  0.313683      3
1486  0.312404      3
1487  0.307287      3
1488  0.306008      3

[1489 rows x 2 columns]


# Random Forest

In [8]:
import pandas as pd
import numpy as np
from scipy.signal import resample
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# Pfad zur Datei
file_path = ['/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S14/S14.pkl',
             '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S13/S13.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S10/S10.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S5/S5.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S7/S7.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S9/S9.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S15/S15.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S2/S2.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S6/S6.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S3/S3.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S4/S4.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S16/S16.pkl',
            '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S17/S17.pkl'
            ]

features = []
window_labels = []

for file in file_path:
# Laden
    with open(file, 'rb') as f:
        data = pd.read_pickle(f)
    
    # Zugriff auf Wrist-Daten
    wrist = data['signal']['wrist']
    eda = wrist['EDA']      # shape (n_samples, 1)
    temp = wrist['TEMP']    # shape (n_samples, 1)
    acc = wrist['ACC']      # shape (n_samples, 3)
    
    # Label (ursprünglich 700 Hz, z. B. 3.8 Mio lang)
    labels = data['label']
    
    # Resampling der Labels auf Länge von EDA (64 Hz)
    labels_resampled = resample(labels.astype(float), len(eda))
    labels_resampled = np.round(labels_resampled).astype(int)
    
    # Features kombinieren (z. B. EDA, TEMP, ACC)
    from scipy.signal import resample
    
    # Resample ACC auf Länge von EDA (z. B. 22192 Zeilen)
    acc_resampled = resample(acc, len(eda))
    
    X_raw = np.hstack([eda, temp, acc_resampled])  # shape: (n_samples, 6)

    X_raw
    
    # Sliding Window-Feature-Bildung (z. B. 60s bei 64 Hz = 384 Samples)
    window_size = 384  # 60s Fenster
    step_size = 384    # ohne Überschneidung

    
    for start in range(0, len(X_raw) - window_size, step_size):
        end = start + window_size
        window = X_raw[start:end]
        label_window = labels_resampled[start:end]
        
        # Nur Fenster mit sinnvollen Labels behalten (z. B. 1 = Ruhe, 2 = Stress)
        if np.all(np.isin(label_window, [1, 2])):
            # Feature-Vektor: Mittelwert + Std jeder Sensor-Spalte
            feature_vector = np.concatenate([window.mean(axis=0), window.std(axis=0)])
            features.append(feature_vector)
            
            # Mehrheits-Label als Klassenlabel
            majority_label = np.bincount(label_window).argmax()
            window_labels.append(majority_label)

# In Arrays umwandeln
X = np.array(features)
y = np.array(window_labels)

print(f"Datensätze: {X.shape}, Klassenverteilung: {np.bincount(y)}")




Datensätze: (227, 10), Klassenverteilung: [  0 149  78]


In [9]:
X_raw[8]

array([  1.968359  ,  33.34      , -49.9914787 , -38.16717886,
         5.37479379])

In [10]:
# Modell trainieren auf 80:20 Split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.2,     # 20% Testdaten
    random_state=42,   # Reproduzierbarkeit
)
clf_testdaten = RandomForestClassifier(n_estimators=105, random_state=42)
clf_testdaten.fit(X_train, y_train)


# Test auf 80:20 Split
y_pred_test = clf_testdaten.predict(X_test)
print(classification_report(y_test, y_pred_test, digits=3))

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import accuracy_score, f1_score
import numpy as np

from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score

def finde_bestes_n_estimators(X_train, X_test, y_train, y_test, 
                              n_min=10, n_max=200, step=10, average="weighted"):
    ergebnisse = {}
    
    for n in range(n_min, n_max+1, step):
        clf = RandomForestClassifier(n_estimators=n, random_state=42)
        clf.fit(X_train, y_train)
        y_pred = clf.predict(X_test)
        
        acc = accuracy_score(y_test, y_pred)
        f1  = f1_score(y_test, y_pred, average=average)
        
        ergebnisse[n] = {"accuracy": acc, "f1_score": f1}
        #print(f"n_estimators={n}: Accuracy={acc:.3f}, F1={f1:.3f}")
    
    # Auswahl nach bestem F1-Score
    best_n = max(ergebnisse, key=lambda k: ergebnisse[k]["f1_score"])
    best_vals = ergebnisse[best_n]
    
    print(f"\n➡️ Bestes n_estimators: {best_n} "
          f"mit Accuracy={best_vals['accuracy']:.3f}, F1={best_vals['f1_score']:.3f}")
    
    return best_n, ergebnisse


finde_bestes_n_estimators(X_train, X_test, y_train, y_test, n_min=10, n_max=200, step=1)


import joblib  # für Speichern/Laden

# Modell trainieren auf kombinierten Trainingsdaten
clf = RandomForestClassifier(n_estimators=100, random_state=42)
clf.fit(X, y)
# ✅ Speichern des gesamten Modells ohne Testdatensplit
joblib.dump(clf, "/kaggle/working/random_forest_model.pkl")


              precision    recall  f1-score   support

           1      0.853     1.000     0.921        29
           2      1.000     0.706     0.828        17

    accuracy                          0.891        46
   macro avg      0.926     0.853     0.874        46
weighted avg      0.907     0.891     0.886        46


➡️ Bestes n_estimators: 15 mit Accuracy=0.913, F1=0.910


['/kaggle/working/random_forest_model.pkl']

## Daten der S8 ziehen und einsetzen!

In [11]:
import pandas as pd
import numpy as np
from scipy.signal import resample
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# 📁 Pfad zur Datei (S12)
file_path_S12 = '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S8/S8.pkl'

# 🔄 Daten laden
with open(file_path_S12, 'rb') as f:
    data_S12 = pd.read_pickle(f)

# 📦 Zugriff auf Wrist-Signale
wrist_S12 = data_S12['signal']['wrist']
eda_S12 = wrist_S12['EDA']
temp_S12 = wrist_S12['TEMP']
acc_S12 = wrist_S12['ACC']

# 🏷️ Labels (ursprünglich 700 Hz)
labels_S12 = data_S12['label']

# 🔁 Resampling der Labels auf EDA-Länge (64 Hz)
labels_resampled_S12 = resample(labels_S12.astype(float), len(eda_S12))
labels_resampled_S12 = np.round(labels_resampled_S12).astype(int)

# 🔁 Resample ACC auf EDA-Länge
acc_resampled_S12 = resample(acc_S12, len(eda_S12))

# 🧩 Features kombinieren: EDA, TEMP, ACC → (n_samples, 6)
X_raw_S12 = np.hstack([eda_S12, temp_S12, acc_resampled_S12])

# 🪟 Sliding-Window Parameter
window_size_S12 = 384  # 60s Fenster
step_size_S12 = 384

features_S12 = []
window_labels_S12 = []

# 🚶 Sliding-Window-Verarbeitung
for start in range(0, len(X_raw_S12) - window_size_S12, step_size_S12):
    end = start + window_size_S12
    window_S12 = X_raw_S12[start:end]
    label_window_S12 = np.round(labels_resampled_S12[start:end]).astype(int)

    unique_labels_S12 = np.unique(label_window_S12)
    
    # ✅ Nur Fenster mit ausschließlich Label 1 oder 2
    if set(unique_labels_S12).issubset({1, 2}):
        majority_label_S12 = np.bincount(label_window_S12).argmax()
        feature_vector_S12 = np.concatenate([window_S12.mean(axis=0), window_S12.std(axis=0)])
        features_S12.append(feature_vector_S12)
        window_labels_S12.append(majority_label_S12)

# 🎯 Nur Label 1 & 2 verwenden
mask_S12 = (np.array(window_labels_S12) == 1) | (np.array(window_labels_S12) == 2)
X_S12 = np.array(features_S12)[mask_S12]
y_S12 = np.array(window_labels_S12)[mask_S12]

# ✅ Sicherheits-Check
assert len(X_S12) == len(y_S12), "Unterschiedliche Anzahl von Features und Labels!"

# ℹ️ Ergebnis anzeigen
print(f"Datensätze: {X_S12.shape}, Klassenverteilung: {np.bincount(y_S12)}")


Datensätze: (17, 10), Klassenverteilung: [ 0 11  6]


In [12]:
print(X_S12.shape, y_S12.shape)
print(X.shape, y.shape)

# ⬇️ Laden
rf_loaded = joblib.load("/kaggle/working/random_forest_model.pkl")

# Testen ob es klappt
print("TEST nach Laden", rf_loaded.predict(X_S12[:5]))

# Test auf Subject S12
y_pred = clf.predict(X_S12)
print(classification_report(y_S12, y_pred, digits=3))

(17, 10) (17,)
(227, 10) (227,)
TEST nach Laden [1 1 1 1 1]
              precision    recall  f1-score   support

           1      0.733     1.000     0.846        11
           2      1.000     0.333     0.500         6

    accuracy                          0.765        17
   macro avg      0.867     0.667     0.673        17
weighted avg      0.827     0.765     0.724        17



In [13]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score

def finde_bestes_n_estimators(X_train, X_test, y_train, y_test, 
                              n_min=10, n_max=200, step=10, average="weighted"):
    ergebnisse = {}
    
    for n in range(n_min, n_max+1, step):
        clf = RandomForestClassifier(n_estimators=n, random_state=42)
        clf.fit(X_train, y_train)
        y_pred = clf.predict(X_test)
        
        acc = accuracy_score(y_test, y_pred)
        f1  = f1_score(y_test, y_pred, average=average)
        
        ergebnisse[n] = {"accuracy": acc, "f1_score": f1}
        #print(f"n_estimators={n}: Accuracy={acc:.3f}, F1={f1:.3f}")
    
    # Auswahl nach bestem F1-Score
    best_n = max(ergebnisse, key=lambda k: ergebnisse[k]["f1_score"])
    best_vals = ergebnisse[best_n]
    
    print(f"\n➡️ Bestes n_estimators: {best_n} "
          f"mit Accuracy={best_vals['accuracy']:.3f}, F1={best_vals['f1_score']:.3f}")
    
    return best_n, ergebnisse

print("Übersicht LOSO nur S8", X_S12.shape, y_S12.shape)
print("Übersicht WeSad der Rest(alle)",X.shape, y.shape)

# Setze X und Y (alle außer S8 ein) und schaue was der beste n_estimator ist
best_estimator = finde_bestes_n_estimators(X, X_S12, y, y_S12, n_min=10, n_max=200, step=1,average="weighted")

# Bestes n_estimators: 27 mit Accuracy=0.941
# Bestes n_estimators: 27 mit Accuracy=0.941, F1=0.940

Übersicht LOSO nur S8 (17, 10) (17,)
Übersicht WeSad der Rest(alle) (227, 10) (227,)

➡️ Bestes n_estimators: 27 mit Accuracy=0.941, F1=0.940


## Alternative zu Random Forest testen

## XGBoost

In [14]:
!pip install xgboost --quiet


In [15]:
import xgboost as xgb
from sklearn.metrics import classification_report
import xgboost as xgb
from sklearn.metrics import classification_report
import numpy as np

# ✅ Filter: Nur Labels 1 (entspannt) & 2 (gestresst)
mask_train = (y == 1) | (y == 2)
mask_test = (y_S12 == 1) | (y_S12 == 2)

X_train_filtered = X[mask_train]
y_train_filtered = y[mask_train]

X_test_filtered = X_S12[mask_test]
y_test_filtered = y_S12[mask_test]


# Optional: Konvertiere in DMatrix (XGBoost-eigenes Format) – ist aber nicht zwingend
# dtrain = xgb.DMatrix(X, label=y)
# dtest = xgb.DMatrix(X_S12, label=y_S12)

# Mapping: 1 → 0 (Entspannt), 2 → 1 (Stress)
y_train_remap = (y_train_filtered - 1).astype(int)
y_test_remap = (y_test_filtered - 1).astype(int)


xgb_clf = xgb.XGBClassifier(
    objective='binary:logistic',  # da nur 2 Klassen nach dem Mapping
    n_estimators=200,
    max_depth=5,
    learning_rate=0.27,
    subsample=0.8,
    colsample_bytree=0.8,
    use_label_encoder=False,
    eval_metric='auc',
    random_state=42
)

xgb_clf.fit(X_train_filtered, y_train_remap)
y_pred = xgb_clf.predict(X_test_filtered)

# Evaluation wieder mit Original-Labels:
from sklearn.metrics import classification_report
print(classification_report(y_test_remap, y_pred, digits=3))


              precision    recall  f1-score   support

           0      0.909     0.909     0.909        11
           1      0.833     0.833     0.833         6

    accuracy                          0.882        17
   macro avg      0.871     0.871     0.871        17
weighted avg      0.882     0.882     0.882        17



-> Gutes Ergebnis:
Alle Personen außer S8; sonst alle drin:

              precision    recall  f1-score   support

           0      0.909     0.909     0.909        11
           1      0.833     0.833     0.833         6

    accuracy                          0.882        17
   macro avg      0.871     0.871     0.871        17
weighted avg      0.882     0.882     0.882        17

-> 5/6 Stressfälle erkannt. 
Lernrate manuell angepasst -> Automatisiert noch besser.

-> Interpretation: XGBoost besser als Random Forest, da Gradiant Boosting: Fehler aus vorherigen Bäumen finden + sich verbessern.

Es kann auch komplexere Zusammenhänge modellieren, wo Random Forest eher Mittelwerte bildet.

Boosting ist oft besser für kleine, saubere Zeitfenster-Daten als vorher.

Settings:
xgb_clf = xgb.XGBClassifier(
    objective='binary:logistic',  # da nur 2 Klassen nach dem Mapping
    n_estimators=200,
    max_depth=5,
    learning_rate=0.27,
    subsample=0.8,
    colsample_bytree=0.8,
    use_label_encoder=False,
    eval_metric='logloss',
    random_state=42
)





## Modell speichern

In [16]:
import joblib

# Modell speichern
joblib.dump(xgb_clf, "/kaggle/working/xgb_stress_model.pkl")

['/kaggle/working/xgb_stress_model.pkl']

## Herausfinden, welche Parameter am besten geeignet sind

In [17]:
import xgboost as xgb
from sklearn.metrics import accuracy_score, f1_score
from sklearn.model_selection import ParameterGrid

def finde_beste_xgb_parameter(X_train, X_test, y_train, y_test, param_grid, average="weighted"):
    ergebnisse = []
    
    for params in ParameterGrid(param_grid):
        clf = xgb.XGBClassifier(
            objective='binary:logistic',
            use_label_encoder=False,
            eval_metric='auc',
            random_state=42,
            **params
        )
        
        clf.fit(X_train, y_train)
        y_pred = clf.predict(X_test)
        
        acc = accuracy_score(y_test, y_pred)
        f1  = f1_score(y_test, y_pred, average=average)
        
        ergebnisse.append((params, acc, f1))
        #print(f"Params={params} | Accuracy={acc:.3f}, F1={f1:.3f}")
    
    # Sortierung nach F1, dann Accuracy
    best_params, best_acc, best_f1 = max(ergebnisse, key=lambda x: (x[2], x[1]))
    
    print(f"\n➡️ Beste Parameter: {best_params}")
    print(f"   Accuracy={best_acc:.3f}, F1={best_f1:.3f}")
    
    return best_params, best_acc, best_f1


In [18]:
X_train_filtered = X[mask_train]
y_train_filtered = y[mask_train]

X_train_filtered = (X_train_filtered - 1).astype(int)
y_train_filtered = (y_train_filtered - 1).astype(int)

X_train, X_test, y_train, y_test = train_test_split(
    X_train_filtered, y_train_filtered, 
    test_size=0.2,     # 20% Testdaten
    random_state=42,   # Reproduzierbarkeit
)
param_grid = {
    "n_estimators": [10, 20, 30, 40, 50, 60,70, 80, 100, 150,175, 200,225, 250, 300],
    "max_depth": [3, 5, 7, 9, 11, 13, 15],
    "learning_rate": [0.05, 0.1,0.15,0.2, 0.25, 0.30],
    "subsample": [0.6,0.7,0.8,0.9, 1.0],
    "colsample_bytree": [0.6, 0.7,0.75, 0.8,0.9,1.0]
}

print("Beste Parameter mit 80 20")
beste_params, beste_acc, beste_f1 = finde_beste_xgb_parameter(
    X_train, X_test, y_train, y_test, param_grid
)


Beste Parameter mit 80 20

➡️ Beste Parameter: {'colsample_bytree': 0.7, 'learning_rate': 0.25, 'max_depth': 5, 'n_estimators': 150, 'subsample': 0.6}
   Accuracy=0.957, F1=0.956


## Testen mit S8 als LOSO Verfahren

In [19]:
X_test = X_S12[mask_test]
y_test = y_S12[mask_test]

X_train, X_test, y_train, y_test = train_test_split(
    X_train_filtered, y_train_filtered, 
    test_size=0.2,     # 20% Testdaten
    random_state=42,   # Reproduzierbarkeit
)
param_grid = {
    "n_estimators": [10, 20, 30, 40, 50, 60,70, 80, 100, 150,175, 200,225, 250, 300],
    "max_depth": [3, 5, 7, 9, 11, 13, 15],
    "learning_rate": [0.05, 0.1,0.15,0.2, 0.25, 0.30],
    "subsample": [0.6,0.7,0.8,0.9, 1.0],
    "colsample_bytree": [0.6, 0.7,0.75, 0.8,0.9,1.0]
}
print("Beste Parameter mit LOSO")

beste_params, beste_acc, beste_f1 = finde_beste_xgb_parameter(
    X_train, X_test, y_train, y_test, param_grid
)


Beste Parameter mit LOSO

➡️ Beste Parameter: {'colsample_bytree': 0.7, 'learning_rate': 0.25, 'max_depth': 5, 'n_estimators': 150, 'subsample': 0.6}
   Accuracy=0.957, F1=0.956


## SVM ("Support Vector Maschines")


In [20]:
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report
import numpy as np

# 🔄 Nur Labels 1 und 2 (wie bei XGBoost)
mask_train = (y == 1) | (y == 2)
mask_test = (y_S12 == 1) | (y_S12 == 2)

X_train_filtered = X[mask_train]
y_train_filtered = y[mask_train]
X_test_filtered = X_S12[mask_test]
y_test_filtered = y_S12[mask_test]

# 🔄 Labels ummappen: 1 → 0, 2 → 1
y_train_remap = (y_train_filtered - 1).astype(int)
y_test_remap = (y_test_filtered - 1).astype(int)

# 🔄 Features skalieren (wichtig für SVM!)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_filtered)
X_test_scaled = scaler.transform(X_test_filtered)

from sklearn.model_selection import GridSearchCV

param_grid = {
    'C': [10,11,12,13,14,15],
    'gamma': [0.01,0.02, 0.03, 0.04, 0.05,0.06, 0.07, 0.08,0.09, 0.1,0.15,0.2,0.3,0.4, 1, 'scale', 'auto'],
    'kernel': ['rbf','linear','polynomial']
}

grid = GridSearchCV(SVC(), param_grid, cv=3, scoring='f1')
grid.fit(X_train_scaled, y_train_remap)

print("Beste Parameter:", grid.best_params_)
print("")

# ✅ SVM mit RBF-Kernel
svm_clf = SVC(kernel='rbf', C=10, gamma=0.1, random_state=42)
svm_clf.fit(X_train_scaled, y_train_remap)

# 🧪 Vorhersage & Bewertung
y_pred = svm_clf.predict(X_test_scaled)
print(classification_report(y_test_remap, y_pred, digits=3))


Beste Parameter: {'C': 12, 'gamma': 0.05, 'kernel': 'rbf'}

              precision    recall  f1-score   support

           0      0.400     0.182     0.250        11
           1      0.250     0.500     0.333         6

    accuracy                          0.294        17
   macro avg      0.325     0.341     0.292        17
weighted avg      0.347     0.294     0.279        17



/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py:378: FitFailedWarning: 
306 fits failed out of a total of 918.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
306 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.11/dist-packages/sklearn/svm/_base.py", line 180, in fit
    self._validate_params()
  File "/usr/local/lib/python3.11/dist-packages/sklearn/base.py", line 600, in _validate_params
    validate_parameter_constraints(
  File "/usr/local/lib/python3.11/dist-packages/sklearn/utils/_pa

In [21]:
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, StratifiedKFold
from sklearn.metrics import f1_score, accuracy_score, make_scorer

def optimiere_svm(X_train, y_train, cv_splits=5, class_imbalance=True):
    """
    Sucht SVM-Hyperparameter mit CV, optimiert primär F1 (weighted),
    nutzt Accuracy als Tie-Breaker und gibt bestes Modell + Ergebnisse zurück.
    """
    pipe = Pipeline([
        ("scaler", StandardScaler()),
        ("svc", SVC(probability=True, class_weight="balanced" if class_imbalance else None))
    ])

    zwischenwerte = [round(x, 2) for x in np.arange(5, 6.01, 0.01)]
    svc_gamma_2 = [round(x, 2) for x in np.arange(0.1, 0.2, 0.001)]
    svc_gamma_2.extend(["scale", "auto"])

    
    param_grid = [
        {"svc__kernel": ["rbf"],
         "svc__C" :zwischenwerte,
         #"svc__C": [0.1, 1, 3,5,5.5, 5.8, 6,6.5, 6.7, 7,7.5, 8, 9, 10,11,13],
         # "svc__gamma": ["scale", "auto", 0.01, 0.03,0.04,0.06,0.08,0.09, 0.1, 0.11, 0.12,0.13]},
         "svc__gamma": svc_gamma_2,
        },
        {"svc__kernel": ["linear"],
         "svc__C": [0.1, 0.01, 0.03,0.04,0.06,0.08,0.09, 0.1, 0.11, 0.12,0.13, 1, 3, 10]},                 # gamma wird ignoriert
        {"svc__kernel": ["poly"],
         "svc__degree": [2, 3],
         "svc__coef0": [0.0, 0.5, 1.0],
         "svc__C": [0.1, 1, 3, 10],
         "svc__gamma": ["scale", "auto", 0.01, 0.03,0.04,0.06,0.08,0.09, 0.1, 0.11, 0.12,0.13]},
    ]

    # Mehrere Metriken loggen; refit = F1 (weighted)
    scorers = {
        "f1": "f1_weighted",
        "accuracy": "accuracy"
    }

    cv = StratifiedKFold(n_splits=cv_splits, shuffle=True, random_state=42)
    grid = GridSearchCV(pipe, param_grid, scoring=scorers, refit="f1",
                        cv=cv, n_jobs=-1, verbose=0, return_train_score=False)
    grid.fit(X_train, y_train)

    # Zusätzlich: Bei gleichem F1 bestes Accuracy-Modell wählen (Tie-Breaker)
    results = grid.cv_results_
    mean_f1 = results["mean_test_f1"]
    mean_acc = results["mean_test_accuracy"]
    best_f1 = mean_f1.max()
    # alle Kandidaten mit maximalem F1 (numerische Toleranz)
    idxs = [i for i, v in enumerate(mean_f1) if abs(v - best_f1) < 1e-9]
    # wähle unter diesen die höchste Accuracy
    best_idx = max(idxs, key=lambda i: mean_acc[i])

    best_params = results["params"][best_idx]
    best_model = grid.best_estimator_ if grid.best_index_ == best_idx else Pipeline([
        ("scaler", StandardScaler())
    ])
    # falls Tie-Breaker anderes Set wählte, neu fitten:
    if grid.best_index_ != best_idx:
        best_model = Pipeline([
            ("scaler", StandardScaler()),
            ("svc", SVC(probability=True,
                       class_weight="balanced" if class_imbalance else None))
        ])
        best_model.set_params(**{k: v for k, v in best_params.items()})
        best_model.fit(X_train, y_train)

    summary = {
        "best_params": best_params,
        "cv_mean_f1": mean_f1[best_idx],
        "cv_mean_accuracy": mean_acc[best_idx]
    }
    return best_model, summary, grid


In [22]:
X_train_filtered = X[mask_train]
y_train_filtered = y[mask_train]
X_test_filtered = X_S12[mask_test]
y_test_filtered = y_S12[mask_test]



best_model, summary, grid = optimiere_svm(X_train_filtered, y_train_filtered, cv_splits=5, class_imbalance=True)
print(summary["best_params"])
print(f"CV F1={summary['cv_mean_f1']:.3f}, CV Acc={summary['cv_mean_accuracy']:.3f}")


{'svc__C': 5.19, 'svc__gamma': 0.12, 'svc__kernel': 'rbf'}
CV F1=0.964, CV Acc=0.965


In [23]:
# Modell speichern
import joblib

# Modell speichern
joblib.dump(svm_clf, "/kaggle/working/svm_clf_stress_model.pkl")

['/kaggle/working/svm_clf_stress_model.pkl']

## CNN 

In [24]:
import numpy as np

def discrete_resample_by_index(labels: np.ndarray, target_len: int) -> np.ndarray:
    """
    Mappt diskrete Labels (z.B. [0,1,2]) ohne Interpolation auf eine neue Länge.
    Verwendet nearest/forward index mapping statt Fourier-Resampling.
    """
    src_len = len(labels)
    if src_len == target_len:
        return labels.astype(int)

    # Indizes im Quellsignal, die jeweils dem EDA-Sample am nächsten sind
    idx = np.floor(np.linspace(0, src_len - 1, target_len)).astype(int)
    return labels[idx].astype(int)

def majority_label_in_window(win_labels: np.ndarray, valid_classes=(1,2), min_valid_ratio=0.8):
    """
    Gibt das Mehrheitslabel im Fenster zurück, wenn genug gültige Labels vorhanden sind,
    sonst None (Fenster verwerfen).
    """
    valid_mask = np.isin(win_labels, valid_classes)
    if valid_mask.mean() < min_valid_ratio:
        return None
    # Nur über gültige Labels die Mehrheit bilden
    counts = np.bincount(win_labels[valid_mask])
    return np.argmax(counts)


In [25]:
import pandas as pd
import numpy as np
from scipy.signal import resample
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

file_paths = [
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S14/S14.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S13/S13.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S10/S10.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S5/S5.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S7/S7.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S9/S9.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S15/S15.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S2/S2.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S6/S6.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S3/S3.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S4/S4.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S16/S16.pkl',
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S17/S17.pkl'
]

# ChatGPT
X_all_list = []
y_all_list = []

def zscore_subject(X):
    mu = X.mean(axis=(0,1), keepdims=True)
    sd = X.std(axis=(0,1), keepdims=True) + 1e-6
    return (X - mu) / sd

for file in file_paths:
    with open(file, 'rb') as f:
        data = pd.read_pickle(f)

    wrist = data['signal']['wrist']
    eda = wrist['EDA']
    temp = wrist['TEMP']
    acc = wrist['ACC']
    labels = data['label']

    # Resample Labels auf EDA-Länge
    #labels_resampled = resample(labels.astype(float), len(eda))
    #labels_resampled = np.round(labels_resampled).astype(int)
    
    # (Optional) Nicht verwendbare Labelcodes auf 0 setzen
    labels_clean = np.array(labels).astype(int)
    valid_classes = {1, 2}
    labels_clean = np.where(np.isin(labels_clean, list(valid_classes)), labels_clean, 0)
    
    # Diskretes Resampling ohne Mischen
    labels_resampled = discrete_resample_by_index(labels_clean, len(eda))
    
    # ACC resamplen und alle Kanäle kombinieren
    acc_resampled = resample(acc, len(eda))
    X_raw = np.hstack([eda, temp, acc_resampled])  # shape: (n_samples, 5)

    # Sliding Window
    #window_size = 384
    #step_size = 384

    #for start in range(0, len(X_raw) - window_size, step_size):
        #end = start + window_size
        #window = X_raw[start:end]
        #label_window = labels_resampled[start:end]

        #if set(np.unique(label_window)).issubset({1, 2}):
            #X_all_list.append(window)
           # majority_label = np.bincount(label_window).argmax()
            #y_all_list.append(majority_label)

    window_size = 384
    step_size = 384
    
    for start in range(0, len(X_raw) - window_size, step_size):
        end = start + window_size
        window = X_raw[start:end]
        label_window = labels_resampled[start:end]
    
        maj = majority_label_in_window(label_window, valid_classes=(1,2), min_valid_ratio=0.8)
        if maj is None:
            continue  # Fenster verwerfen, zu wenig gültige Labels
        X_all_list.append(window)
        y_all_list.append(maj)


X_all = [zscore_subject(X) for X in X_all_list]  # Liste aller Subjekt-Daten

# In Arrays umwandeln
X_all = np.array(X_all_list, dtype=np.float32)          # shape: (n_windows_total, 384, 5)
y_all = (np.array(y_all_list, dtype=np.int64) - 1)      # [1,2] -> [0,1]


print(f"Datensätze: {X_all.shape}, Klassenverteilung: {np.bincount(y_all)}")


Datensätze: (236, 384, 5), Klassenverteilung: [154  82]


In [26]:
#y_cnn = (y_all - 1).astype(int)  # [1, 2] → [0, 1]
X_train, X_test, y_train, y_test = train_test_split(
    X_all, y_all, test_size=0.2, stratify=y_all, random_state=42
)
#X_train

In [27]:
import tensorflow as tf
from tensorflow.keras import layers, models

model = models.Sequential([
    layers.Conv1D(32, kernel_size=6, activation='relu', input_shape=(384, 5)),
    #layers.Conv1D(32, kernel_size=2, activation='relu', input_shape=(384, 5)),
    layers.MaxPooling1D(pool_size=2),
    layers.Conv1D(64, kernel_size=3, activation='relu'),
    layers.MaxPooling1D(pool_size=2),
    #layers.Flatten(),
    layers.GlobalAveragePooling1D(),
    layers.Dense(64, activation='relu'),
    layers.Dropout(0.3),
    layers.Dense(1, activation='sigmoid')  # binary classification
])

model.compile(optimizer='adam', loss='binary_crossentropy',
              metrics=[tf.keras.metrics.BinaryAccuracy(name='accuracy'),
                       tf.keras.metrics.Precision(name='precision'),
                       tf.keras.metrics.Recall(name='recall')])
model.summary()


2025-09-25 14:01:42.100727: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1758808902.352621      13 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1758808902.420961      13 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
2025-09-25 14:01:59.685925: E external/local_xla/xla/stream_executor/cuda/cuda_driver.cc:152

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d (Conv1D)                 │ (None, 379, 32)        │           992 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d (MaxPooling1D)    │ (None, 189, 32)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_1 (Conv1D)               │ (None, 187, 64)        │         6,208 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_1 (MaxPooling1D)  │ (None, 93, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ global_average_pooling1d        │ (None, 64)             │             0 │
│ (GlobalAveragePooling1D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 64)             │         4,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 11,425 (44.63 KB)

 Trainable params: 11,425 (44.63 KB)

 Non-trainable params: 0 (0.00 B)

In [28]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint, ReduceLROnPlateau

# WICHTIG: AUC/Precision/Recall als Metriken loggen (für Callbacks + F1)
model.compile(
    optimizer=tf.keras.optimizers.Adam(3e-4),
    loss='binary_crossentropy',
    metrics=[
        tf.keras.metrics.BinaryAccuracy(name='accuracy'),
        tf.keras.metrics.Precision(name='precision'),
        tf.keras.metrics.Recall(name='recall'),
        tf.keras.metrics.AUC(name='auc')
    ]
)

callbacks = [
    ModelCheckpoint("best.keras", monitor="val_auc", mode="max", save_best_only=True),
    EarlyStopping(monitor="val_auc", mode="max", patience=8, restore_best_weights=True),
    ReduceLROnPlateau(monitor="val_loss", factor=0.5, patience=3, verbose=1, min_lr=1e-6)
]

# Chat GPT Input
from sklearn.utils.class_weight import compute_class_weight

classes = np.unique(y_train)
weights = compute_class_weight('balanced', classes=classes, y=y_train)
class_weights = dict(zip(classes, weights))

history = model.fit(
    X_train, y_train,
    class_weight=class_weights, 
    epochs=50,
    batch_size=32,
    validation_split=0.2,
    callbacks=callbacks,
    verbose=1
)





# --- Beste Epochen ermitteln (aus der EINEN History) ---
hist = history.history
val_acc = np.array(hist["val_accuracy"])
val_prec = np.array(hist["val_precision"])
val_rec  = np.array(hist["val_recall"])
val_auc  = np.array(hist["val_auc"])

# F1 aus Precision/Recall
val_f1 = 2 * val_prec * val_rec / (val_prec + val_rec + 1e-7)

best_epoch_f1  = int(np.argmax(val_f1)) + 1
best_epoch_acc = int(np.argmax(val_acc)) + 1
best_epoch_auc = int(np.argmax(val_auc)) + 1

print(f"Beste Epoche nach F1:  {best_epoch_f1}  | val_f1={val_f1[best_epoch_f1-1]:.4f}")
print(f"Beste Epoche nach Acc: {best_epoch_acc} | val_accuracy={val_acc[best_epoch_acc-1]:.4f}")
print(f"Beste Epoche nach AUC: {best_epoch_auc} | val_auc={val_auc[best_epoch_auc-1]:.4f}")


Epoch 1/50
5/5 ━━━━━━━━━━━━━━━━━━━━ 6s 322ms/step - accuracy: 0.4466 - auc: 0.4541 - loss: 2.2202 - precision: 0.2755 - recall: 0.4184 - val_accuracy: 0.5263 - val_auc: 0.4875 - val_loss: 1.1206 - val_precision: 0.5294 - val_recall: 0.4737 - learning_rate: 3.0000e-04
Epoch 2/50
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 57ms/step - accuracy: 0.4800 - auc: 0.4782 - loss: 1.8697 - precision: 0.3286 - recall: 0.4042 - val_accuracy: 0.6053 - val_auc: 0.6177 - val_loss: 0.7039 - val_precision: 0.7000 - val_recall: 0.3684 - learning_rate: 3.0000e-04
Epoch 3/50
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step - accuracy: 0.5406 - auc: 0.5429 - loss: 1.2760 - precision: 0.3133 - recall: 0.4216 - val_accuracy: 0.7895 - val_auc: 0.8227 - val_loss: 0.5308 - val_precision: 0.7895 - val_recall: 0.7895 - learning_rate: 3.0000e-04
Epoch 4/50
5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 50ms/step - accuracy: 0.6137 - auc: 0.6431 - loss: 1.0907 - precision: 0.3773 - recall: 0.4606 - val_accuracy: 0.8158 - val_auc: 0.9003 - val_loss: 0.4258 - va

In [29]:
model.save("/kaggle/working/cnn_model.keras")


In [30]:
from tensorflow.keras.models import load_model

# Laden (egal ob .keras oder .h5)
cnn_loaded = load_model("/kaggle/working/cnn_model.keras")

# Test
print(cnn_loaded.predict(X_test[:5]))


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 144ms/step
[[0.03224935]
 [0.9776412 ]
 [0.61260754]
 [0.13980828]
 [0.97947955]]


In [31]:
## Download der Daten
from IPython.display import FileLink

FileLink('/kaggle/working/cnn_model.keras')
FileLink('/kaggle/working/svm_clf_stress_model.pkl')


/kaggle/working/svm_clf_stress_model.pkl

In [32]:
import os, zipfile, shutil

# 1) Sicherstellen, dass die Modelle da sind
files_to_zip = [
    "/kaggle/working/cnn_model.keras",
    "/kaggle/working/random_forest_model.pkl",
    "/kaggle/working/svm_clf_stress_model.pkl",
]

missing = [p for p in files_to_zip if not os.path.exists(p)]
if missing:
    print("Fehlende Dateien:", missing)
else:
    print("Alle Dateien vorhanden.")

# 2) Sauberen Export-Ordner anlegen
export_dir = "/kaggle/working/export"
os.makedirs(export_dir, exist_ok=True)

# 3) Dateien in den Export-Ordner kopieren (optional, hält’s übersichtlich)
for p in files_to_zip:
    if os.path.exists(p):
        shutil.copy(p, os.path.join(export_dir, os.path.basename(p)))

# 4) Zip nur mit diesen Dateien bauen
zip_path = "/kaggle/working/models.zip"
with zipfile.ZipFile(zip_path, "w") as zf:
    for name in os.listdir(export_dir):
        zf.write(os.path.join(export_dir, name), arcname=name)

# 5) Kontrolle: auflisten und Größen anzeigen
for name in os.listdir("/kaggle/working"):
    full = os.path.join("/kaggle/working", name)
    if os.path.isfile(full):
        print(f"{name}: {os.path.getsize(full)/1024:.1f} KB")

print("\nFERTIG: models.zip liegt in /kaggle/working und erscheint im Output-Tab.")


Alle Dateien vorhanden.
svm_clf_stress_model.pkl: 9.8 KB
xgb_stress_model.pkl: 167.1 KB
best.keras: 173.2 KB
models.zip: 499.5 KB
cnn_model.keras: 173.2 KB
random_forest_model.pkl: 316.2 KB
__notebook__.ipynb: 133.9 KB

FERTIG: models.zip liegt in /kaggle/working und erscheint im Output-Tab.


## Ausprobieren CNN für "gelernte Daten" 

In [33]:
from sklearn.metrics import classification_report


from tensorflow.keras.models import load_model

# Laden (egal ob .keras oder .h5)
cnn_loaded = load_model("/kaggle/working/cnn_model.keras")


y_pred_prob = cnn_loaded.predict(X_test).flatten()
y_pred = (y_pred_prob > 0.5).astype(int)

print(classification_report(y_test, y_pred, digits=3))


2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 124ms/step
              precision    recall  f1-score   support

           0      0.862     0.806     0.833        31
           1      0.684     0.765     0.722        17

    accuracy                          0.792        48
   macro avg      0.773     0.786     0.778        48
weighted avg      0.799     0.792     0.794        48



## Ausprobieren CNN für "NEUE Daten" -> 8

-> Variablen ändern für S12 bzw. S8 
-> Ausführen
-> 

In [34]:
import pandas as pd
import numpy as np
from scipy.signal import resample

# ---------- Hilfsfunktionen ----------
def discrete_resample_by_index(labels: np.ndarray, target_len: int) -> np.ndarray:
    """Diskretes 'Resampling' per Index-Mapping (keine Interpolation)."""
    labels = np.asarray(labels).astype(np.int64)
    src_len = len(labels)
    if src_len == target_len:
        return labels
    idx = np.floor(np.linspace(0, src_len - 1, target_len)).astype(int)
    return labels[idx]

def majority_label_in_window(win_labels: np.ndarray, valid_classes=(1,2), min_valid_ratio=0.7):
    """Mehrheitslabel im Fenster, nur über gültige Klassen; sonst None."""
    vl = np.asarray(win_labels).astype(np.int64)
    valid_mask = np.isin(vl, valid_classes)
    if valid_mask.mean() < min_valid_ratio:
        return None
    counts = np.bincount(vl[valid_mask])
    return int(np.argmax(counts))

# ---------- Daten laden ----------
file_paths = [
    '/kaggle/input/wesad-wearable-stress-affect-detection-dataset/WESAD/S8/S8.pkl'
]

X_all_list_S8, y_all_list_S8 = [], []

window_size = 384
step_size   = 384
valid_classes = (1, 2)

for file in file_paths:
    with open(file, 'rb') as f:
        data = pd.read_pickle(f)

    wrist = data['signal']['wrist']
    eda   = np.asarray(wrist['EDA']).reshape(-1, 1)   # (N,1)
    temp  = np.asarray(wrist['TEMP']).reshape(-1, 1)  # (N,1)
    acc   = np.asarray(wrist['ACC'])                  # (N,3)

    labels = np.asarray(data['label']).astype(np.int64)

    # 1) Labels reinigen: nur 1/2 behalten, Rest = 0 (ungültig)
    labels_clean = np.where(np.isin(labels, valid_classes), labels, 0)

    # 2) Labels diskret auf EDA-Länge mappen (kein Fourier)
    labels_resampled = discrete_resample_by_index(labels_clean, len(eda))

    # 3) ACC auf EDA-Länge bringen (hier linear/Fourier ok, da kontinuierlich)
    acc_resampled = resample(acc, len(eda))

    # 4) Feature-Matrix bauen
    X_raw = np.hstack([eda, temp, acc_resampled])  # (N,5)

    # 5) Sliding Windows + Majority Voting
    total_windows, kept = 0, 0
    for start in range(0, len(X_raw) - window_size, step_size):
        end = start + window_size
        total_windows += 1

        window = X_raw[start:end]
        label_window = labels_resampled[start:end]

        maj = majority_label_in_window(label_window, valid_classes=valid_classes, min_valid_ratio=0.7)
        if maj is None:
            continue

        X_all_list_S8.append(window)
        y_all_list_S8.append(maj)
        kept += 1

    print(f"S8: Fenster gesamt={total_windows}, behalten={kept}")

# 6) In Arrays umwandeln, Label-Mapping {1,2}->{0,1} genau einmal
X_S8 = np.array(X_all_list_S8, dtype=np.float32)                 # (n_windows, 384, 5)
y_S8 = (np.array(y_all_list_S8, dtype=np.int64) - 1).astype(int) # (n_windows,)

X_S8 = zscore_subject(X_S8)

print(f"Datensätze: {X_S8.shape}, Klassenverteilung: {np.bincount(y_S8) if len(y_S8)>0 else []}")


S8: Fenster gesamt=56, behalten=19
Datensätze: (19, 384, 5), Klassenverteilung: [12  7]


In [35]:
#y_cnn = (y_S8 - 1).astype(int)  # [1, 2] → [0, 1]

#model.fit(X_train, y_train, epochs=20, batch_size=32, validation_split=0.2)


from sklearn.metrics import classification_report

from tensorflow.keras.models import load_model

# Laden (egal ob .keras oder .h5)
cnn_loaded = load_model("/kaggle/working/cnn_model.keras")

y_pred_prob = cnn_loaded.predict(X_S8).flatten()
y_pred = (y_pred_prob > 0.5).astype(int)

print(classification_report(y_S8, y_pred, digits=3))
X_raw[8]

import scipy.ndimage as nd
p_test = cnn_loaded.predict(X_S8, verbose=0).ravel()
p_test = nd.uniform_filter1d(p_test, size=5)   # leichte Glättung



1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 140ms/step
              precision    recall  f1-score   support

           0      0.750     1.000     0.857        12
           1      1.000     0.429     0.600         7

    accuracy                          0.789        19
   macro avg      0.875     0.714     0.729        19
weighted avg      0.842     0.789     0.762        19

